paasta_tools/mesos_tools.py

```diff
@@ -112,28 +113,41 @@ class MesosSlaveConnectionError(Exception):
 
 
 def get_mesos_leader(hostname=MY_HOSTNAME):
-    """Get the current mesos-master leader's hostname. Raise
-    MesosMasterConnectionError if we can't connect.
+    """Get the current mesos-master leader's hostname.
+    Attempts to determine this by using mesos.cli to query ZooKeeper.
+    If this fails, we fallback to following the HTTP redirect
 
-    :param hostname: The hostname to query mesos-master on
+    :param hostname: The hostname to query mesos-master on (only used if we can't use ZK)
     :returns: The current mesos-master hostname"""
-    redirect_url = 'http://%s:%s/redirect' % (hostname, MESOS_MASTER_PORT)
     try:
-        r = requests.get(redirect_url, timeout=10)
-    except requests.exceptions.ConnectionError as e:
-        # Repackage the exception so upstream code can handle this case without
-        # knowing our implementation details.
-        raise MesosMasterConnectionError(repr(e))
-    r.raise_for_status()
-    return re.search('(?<=http://)[0-9a-zA-Z\.\-]+', r.url).group(0)
+        url = master.CURRENT.host
+        hostname = urlparse(url).hostname
+        # This check is necessary, as if we parse a value such as 'localhost:5050',
+        # it won't have a hostname attribute
+        if hostname:
+            host = socket.gethostbyaddr(hostname)[0]
+            fqdn = socket.getfqdn(host)
+            return fqdn
+        else:
+            raise Exception()
+    except:
```

Comment:
Can you add plenty of log.debugs here? if we are falling through to this backup method, I would like to see it during debug, and it will help in all these cases "what does mesos *think* the leader is?"